### PR TITLE
[refactor][OSS] Adding a pytorch parity unit test

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Discussed on https://github.com/pytorch/pytorch/pull/46750#pullrequestreview-563929958, and I'd like to keep parity in between this PR and fairscale's implementation. We already had a parity test including both ShardedDDP and OSS, this adds a dedicated test for OSS only

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/optim/test_oss.py [hunk 1]
```diff
@@ -662,3 +662,74 @@ def test_state_dict_distributed():
...
+
+        ddp_model_single = copy.deepcopy(model)
+        ddp_optimizer = optimizer(ddp_model_single.parameters(), lr=1e-3)
+        ddp_model = DDP(ddp_model_single, device_ids=[rank], broadcast_buffers=True)
+
+        def check_same_model_params():
+            for pg, ddp_pg in zip(sharded_optimizer.param_groups, ddp_optimizer.param_groups):
+                for p, ddp_p in zip(pg["params"], ddp_pg["params"]):
+                    assert torch.allclose(
+                        p, ddp_p, atol=1e-3
```
Comment (min-xu-ai) [inline tests/optim/test_oss.py:693]: note down how is 1e-3 chosen?

File: tests/optim/test_oss.py [hunk 2]
```diff
@@ -662,3 +662,74 @@ def test_state_dict_distributed():
...
+                    ), f"Model parameters differ in between Pytorch optim and OSS {p} {ddp_p}"
+
+            for b, ddp_b in zip(sharded_ddp_model.buffers(), ddp_model.buffers()):
+                assert torch.allclose(b, ddp_b, atol=1e-3), "Model buffers differ in between Pytorch optim and OSS"
```
Comment (min-xu-ai) [inline tests/optim/test_oss.py:697]: same

File: tests/optim/test_oss.py [hunk 3]
```diff
@@ -662,3 +662,74 @@ def test_state_dict_distributed():
...
+                return ddp_loss
+
+            def closure_sharded(input_tensor=input_tensor):
+                sharded_optimizer.zero_grad()
+                sharded_loss = sharded_ddp_model(input_tensor).abs().sum()
+                sharded_loss.backward()
+                return sharded_loss
+
+            _ = ddp_optimizer.step(closure=closure_ddp)
+            _ = sharded_optimizer.step(closure=closure_sharded)
```
Comment (min-xu-ai) [inline tests/optim/test_oss.py:719]: why not check the returned losses?

File: tests/optim/test_oss.py [hunk 4]
```diff
@@ -662,3 +662,74 @@ def test_state_dict_distributed():
...
+        check_optimizer_equivalence(opt)
+
+    dist.destroy_process_group()
+
+
+@skip_if_no_cuda
+@skip_if_single_gpu
+def test_ddp_parity():
+    temp_file_name = tempfile.mkstemp()[1]
+    world_size = torch.cuda.device_count()
```
Comment (min-xu-ai) [inline tests/optim/test_oss.py:733]: need to log the world size in case it fails and we know what was the value in CI?

**Review by min-xu-ai — APPROVED** [review]
LGTM, some nits below

File: tests/optim/test_oss.py (hunk 1 above)
Comment (blefaudeux) [reply tests/optim/test_oss.py:693]: completely arbitrary, I can check with the default actually, you're right

File: tests/optim/test_oss.py (hunk 3 above)
Comment (blefaudeux) [reply tests/optim/test_oss.py:719]: good point, will do !

**Pull request merged**